BUG: Fixes for unicode field names in Python 2

chrisjbillington opened the pull request:
* Allow specifying a single comma-separated unicode string of names in np.rec.fromarrays
* Allow sorting on unicode field names.

This addresses some of the problems raised in issue #2407

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html#writing-the-commit-message
-->

chrisjbillington committed:
Fixes for unicode field names in Python 2

* Allow specifying a single comma-separated unicode string of names in np.rec.fromarrays
* Allow sorting on unicode field names.

This addresses some of the problems raised in issue #2407

chrisjbillington commented:
Without the changes in this pull request, the following fails in Python 2
```python
x = np.zeros(1, dtype=[(u'test', int)])
x.sort(order=u'test')
```
with
```
Traceback (most recent call last):
  File "91.py", line 4, in <module>
    x.sort(order=u'test')
  File "/home/bilbo/anaconda2/lib/python2.7/site-packages/numpy/core/_internal.py", line 312, in _newnames
    raise ValueError("unsupported order value: %s" % (order,))
ValueError: unsupported order value: test
```

and the following

```python
x = np.core.records.fromarrays([[0], [0]], names=u'a,b', formats=u'i4,i4')
```

fails with
```
Traceback (most recent call last):
  File "91.py", line 6, in <module>
    x = np.core.records.fromarrays([[0], [0]], names=u'a,b', formats=u'i4,i4')
  File "/home/bilbo/anaconda2/lib/python2.7/site-packages/numpy/core/records.py", line 610, in fromarrays
    parsed = format_parser(formats, names, titles, aligned, byteorder)
  File "/home/bilbo/anaconda2/lib/python2.7/site-packages/numpy/core/records.py", line 147, in __init__
    self._setfieldnames(names, titles)
  File "/home/bilbo/anaconda2/lib/python2.7/site-packages/numpy/core/records.py", line 180, in _setfieldnames
    raise NameError("illegal input names %s" % repr(names))
NameError: illegal input names u'a,b'
```

Although the fix from pull request #10672 allowed for unicode field names in Python 2, the two above examples would still fail as the strings were explicitly required to be bytestrings before being passed to lower level code.

tylerjereddy commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -4798,6 +4798,15 @@ def test_unicode_field_names(self):
             assert_raises(UnicodeEncodeError, np.dtype, [(('a', nonencodable_name), int)])
 
+            # A single name string provided to fromarrays() is also allowed to be unicode:
+            x = np.core.records.fromarrays([[0], [0]], names=u'a,b', formats=u'i4,i4')
```
Is it worth checking the value / data structure of `x` produced or only that this doesn't raise an exception?

tylerjereddy reviewed (commented):
Just added a minor comment about one of the tests for now.

Not really related, but I'm also noticing that codecov is causing the CI to report a failure because of the patch coverage diff, which I don't think we want (we may need `patch: false` as used in the SciPy config).

Also may be useful to use actual `skipif` decorators from i.e., pytest instead of the plain Python check on version numbers so that the skips are actually reported in the testing summary, but that's an aside here.

chrisjbillington committed:
Use skipif decorator, check constructed array has correct data

chrisjbillington commented:
I've modified the if: else: block that the tests I added were in to use the skipif decorator.
I've also made the test on constructing an array verify the data is correct, but yes, this test is mostly just checking that it doesn't raise an exception.

eric-wieser commented on numpy/core/_internal.py:
```diff
@@ -294,7 +294,7 @@ def _newnames(datatype, order):
     oldnames = datatype.names
     nameslist = list(oldnames)
-    if isinstance(order, str):
+    if isinstance(order, str) or isinstance(order, unicode):
```
nit: Shorter as `isinstance(order, (str, unicode))`

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -4748,55 +4748,69 @@ def test_assign():
+        assert_equal(x['a'][0], 0)
+        assert_equal(x['b'][0], 1)
+
+    @pytest.mark.skipif(sys.version_info[0] >= 3, reason="Not Python 2")
```
Doesnt' this test work fine on python 3 as well?

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -4748,55 +4748,69 @@ def test_assign():
+        # A single name string provided to fromarrays() is also allowed to be unicode:
+        x = np.core.records.fromarrays([[0], [1]], names=u'a,b', formats=u'i4,i4')
+        assert_equal(x['a'][0], 0)
+        assert_equal(x['b'][0], 1)
```
This part of the test should work on both python versions too?

charris commented:
I tagged it for 1.15.1, but that doesn't mean it will go in. I'd like to get 1.15.1 out in a week or so just to clean up the current bugs.

Is it the case that these fixes are not needed for plain old ASCII?

chrisjbillington commented:
Thanks for the comments! I'll make the nit change, and will allow the tests to run that ought to also work on Python 3.

It's correct that these changes are not relevant to the case of using Python 2 with bytestrings as normal, they only affect the case of using unicode strings in Python 2, which mostly comes up when importing `unicode_literals` from `__future__` in order to maintain a Python2/Python 3 compatible codebase.

These unicode strings must be ascii encodable however and so can't actually contain any non-ascii characters, if they do the checks implemented in PR #10672 will throw an error.

So the two pull requests together are just about allowing the field names to be unicode strings in Python 2 (even though they can't contain non-ascii characters), solely for the purposes of easing Python 3 transitions.

chrisjbillington committed:
cleaner isinstance() check, have new tests to run on py3 as well

chrisjbillington committed:
Fix silly mistake in test

chrisjbillington commented:
Changes implemented. The failing builds appear to be unrelated to these changes.

regarding the "backport candidate" tag - I don't believe there is anything to backport to, since unicode field names have only been allowed for Python 2 since 1.15.0.

mattip commented:
"backport candidtate" means this change might be part of a `1.15.1` release. Currently it only targets `master` which will become `1.16`

charris commented:
@chrisjbillington  You don't want this in 1.15.1?

chrisjbillington commented:
@charris I'm sorry, I misunderstood what 'backport candidate' meant. Please disregard.

charris commented:
Thanks @chrisjbillington .
